source/MaterialXGraphEditor/Graph.cpp

```diff
@@ -2505,187 +2515,206 @@ void Graph::addLink(ed::PinId inputPinId, ed::PinId outputPinId)
         return;
     }
 
-    if (inputPin->_connected == false)
+    // Perform kind check
+    bool kindsMatch = (outputPin->_kind == inputPin->_kind);
```

Comment:
Added check that the output and input pins are not the same kinds (eg. not both Output and not both Input).